src/Server/Handler/Request/CallToolHandler.php

```diff
@@ -77,17 +77,19 @@ public function handle(Request $request, SessionInterface $session): Response|Er
             ]);
 
             return new Response($request->getId(), $result);
-        } catch (ToolNotFoundException $e) {
-            $this->logger->error('Tool not found', ['name' => $toolName]);
-
-            return new Error($request->getId(), Error::METHOD_NOT_FOUND, $e->getMessage());
-        } catch (ToolCallException|ExceptionInterface $e) {
+        } catch (ToolCallException $e) {
             $this->logger->error(\sprintf('Error while executing tool "%s": "%s".', $toolName, $e->getMessage()), [
                 'tool' => $toolName,
                 'arguments' => $arguments,
             ]);
 
-            return Error::forInternalError('Error while executing tool', $request->getId());
+            $errorContent = [new TextContent($e->getMessage())];
+
+            return new Response($request->getId(), CallToolResult::error($errorContent));
+        } catch (ToolNotFoundException $e) {
+            $this->logger->error('Tool not found', ['name' => $toolName]);
+
+            return new Error($request->getId(), Error::METHOD_NOT_FOUND, $e->getMessage());
```

Comment:
we could just go with an early exit in line 63 and make it simpler

Reply:
same for the other ones - or we move the not found exceptions into the providers instead of nullable return type.